docs: p2p package overview (#6096)
algorand/go-algorand: network/README-P2P.md

  - mesh thread to maintain `GossipFanout` number of outgoing peers
  - HTTP Server for external HTTP services (block, catchpoints)
  - `OnNetworkAdvance` listener to react on round advancing

A key difference is that `P2PNetwork` uses `go-libp2p-pubsub` for TX message handling.
Upon start it subscribes to `/algo/tx/0.1.0` topic and publishes TX messages as needed.
The `pubsub` library divides message handling into two stages: validation and processing. Based on
the validation result, a message is either discarded or accepted for further
broadcasting to other peers. This necessitates having separate handlers for TX messages
in `TxHandler`, as we must synchronously determine whether a transaction group is valid:
  - can't ignore fast and broadcast later - will be rejected as a seen message
  - can't accept fast to prevent invalid/expired transactions broadcasting

## Major Components

### HTTP Services

`P2PNetwork` uses libp2p's `http` submodule to handle HTTP traffic over libp2p-managed connection.
It is `http.Handler`-compatible so that service handlers are registered the same way as for `WsNetwork`.

### Phonebook and Peerstore and peer classes

> Originally phonebook designed as an address registry holding permanent (`-p` cli option

Reviewer: ```suggestion
Originally phonebook was designed as an address registry holding permanent (`-p` cli option
```

or `phonebook.json` extra configuration file) and dynamic (SRV DNS records) entries.
These entries later can be later retrieved by a peer role
(`PhoneBookEntryRelayRole` or `PhoneBookEntryArchivalRole`).
A new `PeerStore` (built on top of `libp2p.Peerstore`) resembles the original `Phonebook`
by strictly implementing some of its methods and has the remaining `Phonebook`'s methods
with a slightly different signature - `string` vs `peer.AddrInfo` for address representation.

> The main issue with that entries in `PeerStore` are identified by `PeerID`

Reviewer 2: ```suggestion
The main issue is that entries in `PeerStore` are identified by `PeerID`
```

Both P2PNetwork and WsNetwork have an extra level of peer classification on top of two phonebook's
classes: `PeersConnectedOut`, `PeersConnectedIn`, `PeersPhonebookRelays`, `PeersPhonebookArchivalNodes`.
This allows network clients to be more precise on peers set they want to work with. For example,
ledger service wants `PeersPhonebookArchivalNodes`, and transaction syncer - `PeersConnectedOut`.

### wsPeer

Peers are created in `wsStreamHandler` that is called for both incoming and outgoing connections
(and streams). `incoming` flag is set to true for incoming connection.
At the very beginning of the `wsStreamHandler` one byte read/write happens in order to make sure:
  - Stream is operable
  - A placeholder for a handshake where some meta-data can be exchanged

Each peer gets a read channel `handler.readBuffer` where it enqueues incoming messages for routing
to appropriate handler.

Connected peers are maintained as a `wsPeers` map similarly to the `WsNetwork`.
The main difference between `P2PNetwork` and `WsNetwork` is `http.Client`. Because wsPeers operate
over the multiplexed streams in libp2p-managed connection, a plain `http.Client` would not be able

> to connect to p2p HTTP server. This requires `wsPeer` constructed in `P2PNetwork` to have a special

Reviewer 2: ```suggestion
to connect to a p2p HTTP server. This requires the `wsPeer` constructed in `P2PNetwork` to have a special
```

libp2p-streams compatible `http.Client` produced by `MakeHTTPClientWithRateLimit` helper method.

> It implement rate-limiting approach similar to the regular http clients from `WsNetwork`.

Reviewer 2: ```suggestion
It implements a rate-limiting approach similar to the regular http clients from `WsNetwork`.
```

### Broadcaster

`msgBroadcaster` encapsulates a shared broadcasting logic: priority vs bulk messages (and queues),
data preparation, peers retrieving. Broadcast requests eventually hits
`peer.writeNonBlockMsgs` -> `peer.writeLoopSendMsg` -> `conn.WriteMessage`.
See the diagram denoting the broadcast data flow.

> ```mermaid

Reviewer 2: this is cool!

graph LR

    p2pnet[P2PNetwork]
    wsnet[WsNetwork]
    B[broadcaster]

    p2pnet & wsnet -->  B

    subgraph "wsPeer"
        direction LR
        writeNonBlockMsgs
        Conn[conn.WriteMessage]

        subgraph "writeLoop"
            writeLoopSendMsg
        end

        writeNonBlockMsgs --> writeLoop
        writeLoopSendMsg --> Conn
    end

    B --> writeNonBlockMsgs

    Conn --> WMP2P & WMWS

    subgraph "wsPeerConnP2P"
        WMP2P[WriteMessage]
    end

    subgraph "websocket"
        WMWS[WriteMessage]
    end

    subgraph "libp2p"
        stream.Write
    end

    WMP2P --> libp2p
```

### DHT and Capabilities discovery

DHT is controlled by the `EnableDHTProviders` configuration option and the capabilities
exposed by a node. These capabilities include:
  - `archival`: a listening node with `Archival` config flag set
  - `catchpointStoring`: a listening node configured to store catchpoints
  - `gossip`: a listening node with `EnableGossipService` config flag set

When the `P2PNetwork` starts, the node begins advertising its capabilities by running

> a background goroutine. By default, the DHT implementation pulls bootstrap nodes from

Reviewer 2: ```suggestion
a background goroutine. By default, the underlying DHT implementation pulls bootstrap nodes from
```
